File: Depth-first Search/690.employee-importance.java
/*
 * @lc app=leetcode.cn id=690 lang=java
 *
 * [690] 员工的重要性
 */

// @lc code=start
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/
class Solution {
    public int getImportance(List<Employee> employees, int id) {
        Map<Integer, Employee> map = new HashMap<>();
        for(Employee e : employees) {
            map.put(e.id, e);
        }
        return getImportanceHelper(map, id);
    }

    private int getImportanceHelper(Map<Integer, Employee> map, int id) {
        Employee employee = map.get(id);
        for(int subID : employee.subordinates) {
            employee.importance += getImportanceHelper(map, subID);
        }
        return employee.importance;
    }  
}
// @lc code=end

File: Depth-first Search/690.employee-importance.cpp
/*
 * @lc app=leetcode.cn id=690 lang=cpp
 *
 * [690] 员工的重要性
 */

// @lc code=start
/*
// Employee info
class Employee {
public:
    // It's the unique ID of each node.
    // unique id of this employee
    int id;
    // the importance value of this employee
    int importance;
    // the id of direct subordinates
    vector<int> subordinates;
};
*/
class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        for(auto e : employees) {
            if(e->id == id) {
                if(e->subordinates.size() == 0) {
                    return e->importance;
                }
                for(auto subID : e->subordinates) {
                    e->importance += getImportance(employees, subID);
                }
                return e->importance;
            }
        }
        return 0;
    }
};
// @lc code=end

